### PR TITLE
[Processing] first group of vector geometry algotihms

### LINKED ISSUE #1487: [FEATURE][processing] New algorithm to extend lines
Original commit: https://github.com/qgis/QGIS/commit/8dab2cd4d75f4fdf6d10e0741557915e783722ca by nyalldawson

Allows extending linestrings by a set distance at the start
and end of the line

### LINKED ISSUE #1548: [FEATURE][processing] Remove duplicate fill holes algorithm
Original commit: https://github.com/qgis/QGIS/commit/d4c38f9277ffedf78492aa87db8e6609a68b4eb4 by nyalldawson

The Fill_holes script is duplicated by the DeleteHoles algorithm,
and the DeleteHoles algorithm has many advantages:
- it's a full algorithm rather than a script
- uses qgis geometry classes and maintains curves/z/m
- has unit tests
- doesn't rely on non-standard external library (shapely)

(marked as feature for inclusion in release notes)

### LINKED ISSUE #1944: [FEATURE] Port CheckValidity alg to new API
Original commit: https://github.com/qgis/QGIS/commit/8c73bcbcfe2cdccbbf31e389840e67b648f511c1 by nyalldawson

Also make outputs optional from the alg - now you can select
which outputs you need for your model!

### LINKED ISSUE #1993: [FEATURE] Followup addition of set Z/M values algs with Drop Z/M Values algorithm
Original commit: https://github.com/qgis/QGIS/commit/dd0f18355b4bd482b4b868ffdce6c7cb427126b6 by nyalldawson

Allows easy access to drop any z or m values present in a layer (e.g. if
required for compatibility with a database destination, etc)

### LINKED ISSUE #2055: [processing] [needs-docs] Add aggregate algorithm
Original commit: https://github.com/qgis/QGIS/commit/7dbfc5202f5507a958c0b3c74f9c7c4302b3183b by web-flow

[processing] [needs-docs] Add aggregate algorithm

### LINKED ISSUE #2080: [FEATURE] New algorithm for 'collecting' geometries
Original commit: https://github.com/qgis/QGIS/commit/16c4f830b3ff33e82021de2e505b42a87ee554da by nyalldawson

This is basically the equivalent of the dissolve algorithm, but
instead of a dissolving overlapping geometries the geometries
are instead just collected together into a multipart geometry.

It's designed to slot between the 'promote to multipart' algorithm
(which performs no collection of geometries - it just converts
singleparts to multiparts with 1 part) and the more complex
all-encompassing 'aggregate' algorithm.

### PR DESCRIPTION
**Pull request by ghtmtt**
This is the first of 3 PR for the Processing Vector Geometry section. Supersedes #2240

Fix #2055 fix #2080 fix #1993 fix #1944 fix #1548 fix #1487

reviewers are super welcome!

### CODE REVIEW

**Comment by ghtmtt** [comment]
Travis complains (correctly) for a missing anchor. This is due to the splitting of the PR. It should be definitely safe we all 3 PR will be merged

**Comment by ghtmtt** [comment]
Fixes #2055 #2080 #1993 #1944 #1548 #1487

File: source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst [hunk 1]
```diff
@@ -11,525 +11,799 @@ Vector geometry
...
-Console usage
-.............
 
-::
+.. _qgis_drop_mz_values:
 
-  processing.runalg('qgis:dissolve', input, dissolve_all, field, output)
+Drop m/z values
+---------------
+Removes any measure (M) or Z values from input geometries.
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst]: I like more the second option :)

File: source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst [hunk 2]
```diff
@@ -11,525 +11,799 @@ Vector geometry
...
+.. _qgis_eliminate_selected_polygons:
+
+Eliminate selected polygons
+---------------------------
+Combines selected polygons of the input layer with certain adjacent polygons by
+erasing their common boundary. The adjacent polygon can be either the one with
+the largest or smallest area or the one sharing the largest common boundary with
+the polygon to be eliminated.
+
+The selected features will always be eliminated whether the option "Use only selected features"
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/user_manual/processing_algs/qgis/vector_geometry_tools.rst]: right.

**Comment by ghtmtt** [comment]
@DelazJ I merged manually the content of #2256 #2257 with your review comments. Having a single (reviewed) PR allowed Travis to build correctly so we are sure nothing is wrong. Even if we catch some other errors we can know edit directly this file.

I'm going to merge this so the next week we can check other possible issues between the help system and the docs.

Another huge thanks for all your reviews in this months!

**Comment by DelazJ** [comment]
Issue with merging this way all the PRs in one is that we need to manually close all the concerned issues (+PRs btw). Also it could be nice imho to push changes to those PRs also to help us easily understand how you address some of the comments (anyway, it's more me being curious than a need for doc team).
> Another huge thanks for all your reviews in this months!

You are welcome, dude!
With this commit, do you mean that all qgis algorithms documentation are up to date? I can't find eg, modeler group (from memory given that i currently have a broken Processing on Windows and can't check further). Anyway a cleanup of the repo from processing issues will also be warmly welcomed

**Comment by ghtmtt** [comment]
> Issue with merging this way all the PRs in one is that we need to manually close all the concerned issues (+PRs btw). Also it could be nice imho to push changes to those PRs also to help us easily understand how you address some of the comments (anyway, it's more me being curious than a need for doc team).

yes I know, but even closing related issues and PR manually is better IMHO than a compilation failing. Anyway, I'll handle this in some way ;)

> You are welcome, dude!

No really, without your huge effort all the workflow would have been slower. So thanks!

> With this commit, do you mean that all qgis algorithms documentation are up to date? I can't find eg, modeler group (from memory given that i currently have a broken Processing on Windows and can't check further). Anyway a cleanup of the repo from processing issues will also be warmly welcomed

yes, is the last commit so that all Processing algorithms are updated. For sure they can be enhanced and, as you said in the mailing list, the documentation process is not *fixed*

**Comment by DelazJ** [comment]
> Anyway, I'll handle this in some way ;)

Thanks for having fled our box with this :-). Nice to see list of issues decrease so quickly. 
What i tried to mean in my last comment is that:
1. There's a modeler-only group that no longer exists in docs (hence i guess its tools are not documented. Shouldn't they?)
1. There are also some (few) new/improved/removed(?) tools after you submitted PR for their group. Not sure whether they are currently handled. 
1. There still are open issues that you already address (see for example issues at the bottom of this page https://github.com/qgis/QGIS-Documentation/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+processing I'm pretty sure those are already tackled. Sorry to ask you that but you are currently the most versed in what processing documentation contains: can you give a look to the list and clean it a bit please?

In other words, my main concern is to ensure that the issue repo reflects the reality of the great work done (and, with the hackfest avoid people trying to tackle already fixed issues).
